### ISSUE TITLE
Proximity sensor is not working

### ISSUE THREAD

**Issue by divis1969** [divis1969/android_device_meizu_meilan2#35]
It looks like the sensor is not calibrated properly on some devices

**Comment by divis1969** [comment]
I'm currently stuck with a sensor calibration.
It looks like calibration is provided by libhwmon.so using few ioctls for /dev/als_ps.
Calibration is successful by running MTK enginnering app and consists of 2 stages:
1. Storing some values to /data/nvram/APCFG/APRDCL/HWMON_PS file
2. Applying these values via ioctl

Calibration on boot always fails.
nvram_daemon reads the values from HWMON_PS and tries to ioctl which fails with ENOTTY, which most likely means "request is not supported".

It is interesting I had created my own "daemon" which attempts to ioctl (the same request but zeored data) while starting the phone. It does this before and after the nvram_daemon attempt and it always succeeded.

Thus, I cannot yet guess what could be the reason for ioctl.

The driver is probably similar to the one of p8000 driver (https://github.com/visi0nary/android_kernel_elephone_p8000/blob/master/drivers/misc/mediatek/alsps/alsps.c) and a specific back-end driver is similar to https://github.com/visi0nary/android_kernel_elephone_p8000/blob/master/drivers/misc/mediatek/alsps/epl2182/epl2182.c
But also note that I did not find some kernel logs in a code, so I could be wrongly identified the code.

Perhaps I need to build the kernel from sources to continue debugging.

There is also possibility I'm wrong about the calibration procedure and the components involved.
For example, 'factory' executable also refers to /dev/als_ps and hwmon lib calibration function, but I'm assuming this executable is never run on the production device.

**Comment by agent-007** [comment]
I think, you should use **strace** for debugging ioctl syscals from MTK enginnering app and nvram_daemon. Long time ago i used [this strace binary](http://elinux.org/Android_Tools#strace), but i don't know how this strace works on current android versions.

**Comment by divis1969** [comment]
I've found this ioctl and HWMON_PS exactly from a strace log.
The problem now is I do not know why ioctl of nvram_daemon fails.
Another way to continue investigation is to run strace on stock (or maybe simply collect logs as early as possible)

**Comment by agent-007** [comment]
Did you have full strace (strace -fF -s255 -r) logs from both nvram_daemon and MTK app  ?

**Comment by divis1969** [comment]
I've used -f only, I suppose this is enough to make some conclusion.

Here is the quote:

```
528   openat(AT_FDCWD, "/dev/als_ps", O_RDONLY|O_LARGEFILE <unfinished ...>
528   <... openat resumed> )            = 4
528   ioctl(4, 0x40048411 <unfinished ...>
528   <... ioctl resumed> , 0xfffffffff6603d5c) = -1 ENOTTY (Not a typewriter)
```

**Comment by divis1969** [comment]
I've found one more Meizu/Flyme spacific way to restore calibration on boot. This is implemented by DeviceControlManager class from Flyme.
I've ported it to CM (https://github.com/divis1969/android_frameworks_base/commit/37c4782147aa38b397fd22a954cbf3df481b1f77) and there were few confirmations from users this fixed the issue (on the devices that were suffering from proximity calibration).

So, this issue seems fixed now
